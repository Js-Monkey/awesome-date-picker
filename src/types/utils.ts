export interface UtilObject {
  [key: string]: any
}

export interface UtilFn extends UtilObject {
  (...arg: any): any
}

export interface Types {
  Number: string
  String: string
  Function: string
  Array: string
  Date: string
  Object: string
}

export interface Rect<T = number> {
  width: T
  height: T
  left: T
  top: T
}

export interface Transform<T = string> {
  top: T
  left: T
  bottom: T
  right: T
}

export type eventType = 'click' | 'mouseenter' | 'mouseleave' | 'focus'

export type eventHandler = (e: Event) => unknown

export interface EventListener {
  name: eventType
  handler: eventHandler
}

export interface Style {
  float?: 'left' | 'right'
  color?: string
  backgroundColor?: string
  margin?: string
  'margin-left'?: string
  'margin-right'?: string
  'margin-top'?: string
  'margin-bottom'?: string
  padding?: string
}

export interface CreateElementOptions {
  name?: 'svg' | 'span' | 'div' | 'ul' | 'li' | 'input'
  text?: string
  event?: eventHandler | EventListener[]
  class?: string[]
  style?: Style
  children?: (CreateElementOptions | HTMLElement)[]
  initial?: 'hidden'
}

export interface Handler<T = (el: HTMLElement | Element, options: CreateElementOptions) => void> {
  event: T
  class: T
  style: T
  children: T
  name: () => void
  text: T
  initial: T
}
