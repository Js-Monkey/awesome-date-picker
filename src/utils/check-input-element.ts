import {getNodeName} from './window'

export function isInputElement(el: HTMLElement | null): boolean {
  return getNodeName(el) === 'input'
}
