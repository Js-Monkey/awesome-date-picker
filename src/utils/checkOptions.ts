import Options, {AcceptOptions} from '../types/options'
import {isArray, isBoolean, isFunc, isNumber, isString} from './typeOf'

const acceptOptions: AcceptOptions = {
  placement: ['top', 'left', 'bottom', 'right'],
  type: ['date', 'date-range', 'month', 'month-range'],
  unlinkPanels: (val: unknown) => isBoolean(val),
  format: (val: unknown) => isString(val),
  zIndex: (val: unknown) => isNumber(val),
  disabled: (val: unknown) => isBoolean(val),
  disabledDate: (val: unknown) => isFunc(val),
  offset: (val: unknown) => isNumber(val),
  insertTo: ['body', 'parent'],
  binding: (val: unknown) => isBoolean(val)
}

export function checkOptions(key: keyof Options, val: unknown): boolean {
  const optionVal = acceptOptions[key] as any
  if (isArray(optionVal)) {
    return optionVal.indexOf(val) > -1
  } else {
    return optionVal(val)
  }
}
