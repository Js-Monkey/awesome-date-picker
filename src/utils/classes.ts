const logo = 'better-datepicker'
export const wrapper = `${logo}-wrapper`
export const header = `${logo}-header`
export const day = `${logo}-day`
export const dayBar = `${day}-bar`
export const dayContent = `${day}-content`
export const month = `${logo}-month`
export const year = `${logo}-year`
export const show = `${logo}-show`
export const hidden = `${logo}-hidden`
export const float = `${logo}-float`
