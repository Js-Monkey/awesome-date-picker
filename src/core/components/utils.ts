import {pageName, State} from '../../types/store'
import {isNumber} from '../../utils/typeOf'

export function nextYear(state: State): void {
  state.startYear += 1
}

export function preYear(state: State): void {
  state.startYear -= 1
}

export function nextMonth(state: State): void {
  state.startMonth += 1
}

export function preMonth(state: State): void {
  state.startMonth -= 1
}

export function toMonthPage(state: State, idx: number): void {
  state.page = 'month'
  if (isNumber(idx)) state.startYear += idx
}

export function toDayPage(state: State, month: number): void {
  state.startMonth = month
  state.page = 'day'
}

export function toYearPage(state: State): void {
  state.page = 'year'
}

export function selectDate(day: number): void {
  //todo
}

export function toggleVisibility(page: pageName): string {
  return page === 'day' ? 'show' : 'hidden'
}
