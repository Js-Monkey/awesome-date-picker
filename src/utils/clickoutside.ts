import {_Event} from '../types/event'
import {State} from '../types/store'

export default function clickOutside(state: State, e: _Event): void {
  const {visible, reference, popover} = state
  if (!visible || reference?.contains(e.target as any) || popover?.contains(e.target as any)) return
  state.visible = false
}
