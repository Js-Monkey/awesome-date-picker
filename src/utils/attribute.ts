import {Style, UtilObject} from '../types/utils'
import {isObject} from './typeOf'

export function resetAttr(el: HTMLElement | Element, val: string, name = 'class'): void {
  if (!val) return el.removeAttribute(name)
  el.setAttribute(name, val)
}

export function transformStyle(sty: Style): string {
  return Object.keys(sty)
    .reduce((acc, key) => acc.concat(`${key}:${sty[key as never]}`), [] as string[])
    .join(';')
}

export function addAttr(el: HTMLElement | Element, val: string | UtilObject, name = 'class'): void {
  const attr = el.getAttribute(name)
  if (isObject(val)) {
    val = Object.keys(val).reduce((c, key) => c + key + ':' + val[key as keyof void] + ';', '')
  }
  if (attr && attr.indexOf(val) === -1) val += ' ' + attr
  el.setAttribute(name, val)
}
