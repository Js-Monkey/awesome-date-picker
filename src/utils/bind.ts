import {Fn} from '../types/utils'

export function Bind(fn: Fn, ...arg: any): Fn {
  function proxyFn() {
    fn(...Array.from(arguments).concat(arg))
  }

  return proxyFn
}
