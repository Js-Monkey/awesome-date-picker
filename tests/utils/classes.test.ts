import * as classesName from '../../src/utils/classes'

describe('classes', () => {
  test('className', () => {
    expect(classesName).toMatchSnapshot()
  })
})
