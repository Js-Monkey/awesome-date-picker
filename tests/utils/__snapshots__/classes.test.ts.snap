// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`classes className 1`] = `
Object {
  "day": "better-datepicker-day",
  "defaultCursor": "better-datepicker-default-cursor ",
  "header": "better-datepicker-header",
  "month": "better-datepicker-month",
  "today": "better-datepicker-today",
  "wrapper": "better-datepicker-wrapper",
  "year": "better-datepicker-year",
}
`;
